### PR TITLE
Run `nixosTests` on darwin

### PR DESCRIPTION
**Pull request by roberth**
###### Description of changes

Towards #108984

I had to add `lazyFunction` in order to half-support `make-test-python.nix` calls which don't pass `hostPkgs` along.
All tests should be migrated to use the simpler `runTest` IMO.

Draft because this needs testing and validation, but I don't have much time to spend on this. Figured I'd dump my knowledge in the form of a draft PR (and because I've changed things #191540).

TODO
 - [ ] Eval on hydra against eval regressions
   - [ ] in existing nixos.test.*
   - [ ] future breakage in the macos tests (but only `nixos.tests.simple` and test-framework testing tests?)
 - [ ] Docs
 - [ ] Add tests to (TBD) `nixosTests.nixos-test-runner` suite (categorize tests with `recurseIntoAttrs`)
 - [ ] Consider moving more of the logic into `nixos/lib/testing`. The `nixosTests` system logic is also useful there.
 - [ ] Get #194514 done and rebase
 
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by YorikSar — CHANGES_REQUESTED** [review]
Thanks a lot for tackling this! I couldn't find all ends where `hostPkgs` needs to be pulled through.

I've tried running `nix build -L .#nixosTests.doas` with `--system aarch64-darwin` and `--system x86_64-darwin`. It worked fine with `aarch64` and expectedly broke on `x86_64`. I'm on M1 Mac and QEMU command line currently doesn't support running non-native architecture (needs some tweaks, if that's possible). Would it be possible to warn user about this somehow?

I've also tried running `boot.uefiUsb`, but Nix segfaults on evaluation on my machine for some reason.

Would it be possible to add some of these tests on Hydra, so that we can be sure that this will not get broken in the future?

File: nixos/lib/testing/run.nix [hunk 1]
```diff
@@ -1,6 +1,13 @@
-{ config, hostPkgs, lib, ... }:
+{ config, hostPkgs, pkgs, lib, ... }:
 let
   inherit (lib) types mkOption mdDoc;
+
+  vmHostPlatform = hostPkgs.stdenv.buildPlatform;
+
+  virtualisationFeature =
+    if vmHostPlatform.isLinux then [ "kvm" ]
+    else if vmHostPlatform.isDarwin then [ "hvf" ]
```
[the remark this answers is not in the document]
Comment (roberth) [reply nixos/lib/testing/run.nix:9]: I know some people virtualize macOS on apple hardware in order to make their build farm setup reproducible.
It seems unlikely that nested virtualization would be available on such VMs.

**Comment by roberth** [comment]
I've removed the cross compilation changes, which I wasn't confident about. However I did leave the preliminary support for running NixOS tests using cross-compiled pkgs in place.

**Comment by roberth** [comment]
> I'm on M1 Mac and QEMU command line currently doesn't support running non-native architecture (needs some tweaks, if that's possible). Would it be possible to warn user about this somehow?

I don't know about what can be tweaked and I have no ambition to research this (sorry). A warning could be added if we're confident that this will never be supported, but until we're confident I don't want to discourage anyone from picking up where I/we left off.

> Would it be possible to add some of these tests on Hydra, so that we can be sure that this will not get broken in the future?

I would think so and this is on the todo list. :+1:

**Comment by Gabriella439** [comment]
Is it possible to use `nixosTest` with this branch?  I tried the following:

```nix
{ inputs.nixpkgs.url =
    "github:NixOS/nixpkgs/e39a5efc4504099194032dfabdf60a0c4c78f181";

  outputs = { nixpkgs, ... }: {
    checks.aarch64-darwin.default =
      nixpkgs.legacyPackages.aarch64-darwin.nixosTest {
        name = "test";

        nodes.machine.virtualisation.host.pkgs =
          nixpkgs.legacyPackages.aarch64-darwin;

        testScript = "";
      };
  };
}
```

… but that fails with:

```
error: assertion '(stdenv).isLinux' failed

       at /nix/store/3zj9cv9pj7q2y4g0q1bfxzmv4wq2i4zj-source/pkgs/os-specific/linux/kernel/generic.nix:71:1:

           70|
           71| assert stdenv.isLinux;
             | ^
           72|
(use '--show-trace' to show detailed location information)
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -139,16 +139,35 @@ with pkgs;
...
+        # In the context of a normal darwin compile, repurpose nixosTests as a
+        # non-cross build. This requires a "remote" builder such as a local
+        # Linux VM, but all VM hosting happens on darwin.
+        let linuxSystem = "${stdenv.buildPlatform.parsed.cpu.name}-linux";
+        in {
+          hostPkgs = pkgs;
+          pkgs = import ../.. { inherit config overlays; system = linuxSystem; };
+          system = linuxSystem;
+        }
+      else throw "Don't know how to configure NixOS for VM host ${stdenv.buildPlatform.system} and/or VM guest ${stdenv.hostPlatform.system} built on ${stdenv.buildPlatform.system}";
```
Comment (roberth) [inline pkgs/top-level/all-packages.nix:161]: This logic is useful in other places, so it should be factored or moved to a common location, such as the nixos/lib/testing/pkgs.nix module.

**Comment by roberth** [comment]
> Is it possible to use `nixosTest` with this branch?

It doesn't have a good reason not to, but the immediate problem is that doesn't let you set anything related to either `pkgs` or `hostPkgs`.

It should probably the logic I've highlighted in my previous comment in one way or another.

In your example, perhaps you might get away with setting `nodes.machine.nixpkgs.pkgs` to a linux one by hand?
Really it should be automated though, like `nixosTests` plural.

**Comment by Gabriella439** [comment]
That works!  The following modified version built for me:

```nix
{ inputs.nixpkgs.url =
    "github:NixOS/nixpkgs/e39a5efc4504099194032dfabdf60a0c4c78f181";

  outputs = { nixpkgs, ... }: {
    checks.aarch64-darwin.default =
      nixpkgs.legacyPackages.aarch64-darwin.nixosTest {
        name = "test";

        nodes.machine = {
          nixpkgs.pkgs = nixpkgs.legacyPackages.aarch64-linux.pkgs;

          virtualisation.host.pkgs =
            nixpkgs.legacyPackages.aarch64-darwin;
        };

        testScript = "";
      };
  };
}
```

**Comment by roberth** [comment]
Oh and `nixosTest` should be superseded by a new function that takes a module argument instead of behaving like `callPackage`. Sadly those behaviors are mutually exclusive.
I did add [`nixpkgsFlake.lib.nixos.runTest`](https://nixos.org/manual/nixos/unstable/index.html#sec-call-nixos-test-outside-nixos), which takes a module, but leaves the `pkgs` and `hostPkgs` for you to set. Not exactly the answer to your question, but perhaps a solution?

**Comment by goodlyrottenapple** [comment]
@Gabriella439 I tried building your example flake on an M1 but getting the following error:

```bash
nix build .#checks.aarch64-darwin.default                               
error: a 'aarch64-darwin' with features {hvf, nixos-test} is required to build '/nix/store/icqc6g8lf1f1lnl1dzipfaw5rlmp4vsx-vm-test-run-test.drv', but I am a 'aarch64-darwin' with features {benchmark, big-parallel, nixos-test}
```
(currently on macOS 12.4)

Is there any other setup necessary to get `hvf` working?

**Comment by Gabriella439** [comment]
Oh yeah, I had to also add this to my `nix.conf`:

```
extra-system-features = hvf
```

**Comment by goodlyrottenapple** [comment]
> Oh yeah, I had to also add this to my `nix.conf`:
> 
> ```
> extra-system-features = hvf
> ```

Ah thanks so much! I did try this before, but the option was being ignored when put in `~/.config/nix/nix.conf`. After I moved it to `/etc/nix/nix.conf`, it built successfully.

**Comment by roberth** [comment]
Created #7619 to mirror the solution we have for `kvm` on linux.

**Comment by roberth** [comment]
- Created https://github.com/NixOS/nix/pull/7619 to mirror the solution we have for `kvm` on linux.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (roberth) [reply pkgs/top-level/all-packages.nix:161]: How about `pkgsLinux` (in `pkgs`)

**Pull request closed without merging**